### PR TITLE
fix: capture error stacks across async calls

### LINKED ISSUE #1080: Invalid argument error now explaining what the error is
#### Environment details

  - OS: 10.15.5 Beta (19F83c)
  - Node.js version: 13.7.0
  - npm version: 6.14.4
  - `@google-cloud/firestore` version:  3.7.5

#### Steps to reproduce

I am not sure how to reproduce, all I know is somewhere I am getting:
```
(node:41601) UnhandledPromiseRejectionWarning: Error: 3 INVALID_ARGUMENT: Request contains an invalid argument.
    at Object.callErrorFromStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/call.ts:81:24)
    at Object.onReceiveStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/client.ts:324:36)
    at Object.onReceiveStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/client-interceptors.ts:439:34)
    at Object.onReceiveStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/client-interceptors.ts:402:48)
    at Http2CallStream.outputStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/call-stream.ts:228:22)
    at Http2CallStream.maybeOutputStatus (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/call-stream.ts:278:14)
    at Http2CallStream.endCall (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/call-stream.ts:262:12)
    at Http2CallStream.handleTrailers (/Users/jakeleventhal/myproj/node_modules/@grpc/grpc-js/src/call-stream.ts:398:10)
    at ClientHttp2Stream.emit (events.js:321:20)
    at ClientHttp2Stream.EventEmitter.emit (domain.js:485:12)
    at emit (internal/http2/core.js:289:8)
    at processTicksAndRejections (internal/process/task_queues.js:87:22)
```

The issues are:
1. The error should include what the invalid argument is
2. The error should include the place in the code that the error occurred in the stack trace

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes #1080.

Tested for get/set operations on references, queries, batches, and bulkCommit.

### CODE REVIEW

File: dev/src/bulk-writer.ts [hunk 1]
```diff
@@ -188,7 +188,13 @@ class BulkCommitBatch {
       'The batch should be marked as READY_TO_SEND before committing'
     );
     this.state = BatchState.SENT;
-    return this.writeBatch.bulkCommit();
+
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack;
```
Comment (schmidt-sebastian) [inline dev/src/bulk-writer.ts:193]: Nit: Move bang here.

File: dev/src/index.ts [hunk 2]
```diff
@@ -908,9 +908,15 @@ export class Firestore {
       documentRefsOrReadOptions
     );
     const tag = requestTag();
+
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack;
```
Comment (schmidt-sebastian) [inline dev/src/index.ts:913]: Nit: Move bang here.

File: dev/src/pool.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@
 import * as assert from 'assert';
 
 import {logger} from './logger';
-import {Deferred} from './util';
+import {Deferred, wrapError} from './util';
```
Comment (schmidt-sebastian) [inline dev/src/pool.ts:20]: Unused import?

File: dev/src/reference.ts [hunk 4]
```diff
@@ -1808,12 +1808,15 @@ export class Query<T = DocumentData> {
   _get(transactionId?: Uint8Array): Promise<QuerySnapshot<T>> {
     const docs: Array<QueryDocumentSnapshot<T>> = [];
 
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack;
```
Comment (schmidt-sebastian) [inline dev/src/reference.ts:1812]: Nit: Move bang here.

File: dev/src/write-batch.ts [hunk 5]
```diff
@@ -532,7 +532,13 @@ export class WriteBatch {
    * });
    */
   commit(): Promise<WriteResult[]> {
-    return this.commit_();
+
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts:535]: Nit: Remove empty line.

File: dev/src/write-batch.ts [hunk 6]
```diff
@@ -532,7 +532,13 @@ export class WriteBatch {
...
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack;
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts:537]: Nit: Move bang here.

File: dev/system-test/firestore.ts [hunk 7]
```diff
@@ -2148,6 +2148,17 @@ describe('WriteBatch class', () => {
       });
   });
 
+  it('has a full stack trace if set() errors', () => {
+    // Use an invalid document name that the backend will reject.
+    const ref = randomCol.doc('__doc__');
+    const batch = firestore.batch();
+    batch.set(ref, {foo: 'a'});
+    return batch.commit().then(() => Promise.reject('commit() should have failed')
```
Comment (schmidt-sebastian) [inline dev/system-test/firestore.ts:2156]: I am glad this is causing a lint failure :) Doesn't look pretty like this. The PR is good to go after `yarn gts fix`.

Comment (thebrianchen) [reply dev/system-test/firestore.ts]: fixed. Also I changed `wrapError()` to use `GoogleError` in order to preserve error codes when re-throwing.

File: dev/src/util.ts [hunk 8]
```diff
@@ -146,3 +146,17 @@ export function isPermanentRpcError(
...
+/**
+ * Wraps the provided error in a new error that includes the provided stack.
+ *
+ * Used to preserve stack traces across async calls.
+ * @private
+ */
+export function wrapError(err: GoogleError, stack: string): Error {
+  const wrappedError = new GoogleError(err.message);
+  wrappedError.code = err.code;
+  wrappedError.stack += '\nCaused by: ' + stack;
```
Comment (schmidt-sebastian) [inline dev/src/util.ts:159]: Can we modify the stack in 'err' directly?

Comment (thebrianchen) [reply dev/src/util.ts]: good point. done.

**Pull request merged**